presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -221,38 +227,105 @@ public static HiveSplitSource bucketed(
+
+                        BucketedSplits splits = buckets.computeIfAbsent(bucketNumber.getAsInt(), ignored -> {
                             isNew.set(true);
-                            return new AsyncQueue<>(estimatedOutstandingSplitsPerBucket, executor);
+                            return new BucketedSplits(estimatedOutstandingSplitsPerBucket, executor);
                         });
+
                         if (isNew.get() && finished.get()) {
-                            // Check `finished` and invoke `queue.finish` after the `queue` is added to the map.
-                            // Otherwise, `queue.finish` may not be invoked if `finished` is set while the lambda above is being evaluated.
+                            // Check `finished` and invoke `splits.finish` after the `splits` is added to the map.
+                            // Otherwise, `splits.finish` may not be invoked if `finished` is set while the lambda above is being evaluated.
+                            splits.finish();
+                        }
+                        return splits;
+                    }
+
+                    // This wrapper provides per-bucket level synchronization to support rewind.
+                    // TODO: Rewrite split source for bucketed mode so we don't have two layers of wrappers. (PerBucket -> BucketedSplits -> AsyncQueue)
+                    class BucketedSplits
+                    {
+                        private final int estimatedQueueSize;
+                        private final Executor executor;
+                        private final List<InternalHiveSplit> discoveredHiveSplits = new ArrayList<>();
+
+                        private AsyncQueue<InternalHiveSplit> queue;
+                        private ListenableFuture<List<ConnectorSplit>> lastBorrowBatchAsyncFuture;
+
+                        public BucketedSplits(int estimatedQueueSize, Executor executor)
+                        {
+                            this.estimatedQueueSize = estimatedQueueSize;
+                            this.executor = requireNonNull(executor, "executor is null");
+
+                            this.queue = new AsyncQueue<>(estimatedQueueSize, executor);
+                        }
+
+                        public synchronized ListenableFuture<?> offer(InternalHiveSplit internalHiveSplit)
+                        {
+                            discoveredHiveSplits.add(internalHiveSplit);
+                            return queue.offer(internalHiveSplit);
+                        }
+
+                        public synchronized ListenableFuture<List<ConnectorSplit>> borrowBatchAsync(int maxSize, Function<List<InternalHiveSplit>, BorrowResult<InternalHiveSplit, List<ConnectorSplit>>> function)
+                        {
+                            checkState(lastBorrowBatchAsyncFuture == null || lastBorrowBatchAsyncFuture.isDone(), "last borrowBatchAsync must finish if exists");
```

Comment:
In theory it should always be done. As we are waiting till the end of split enumeration. This is quite confusing.